### PR TITLE
Use scheme for install rules

### PR DESCRIPTION
**Pull request by aalekseyev**
- Make build system explicitly ask for install rules the same way it asks for normal build rules, so that the rules can be computed on demand rather than upfront
- Implement install rules using `Scheme.t`

### CODE REVIEW

**Comment by diml** [comment]
@aalekseyev the patch itself looks good, however looking at the final state of `install_rules.ml`, I feel like we could now simplify things a little. What this module does is:

- compute a list of installation entries by scanning all the stanzas in the workspace
- then, using this list, independently:
  + compute a map of `file -> package name`
  + generate the install file
  + generate symlinks from `_build/<context>` to `_build/install/<context>`

The code is not entirely written this way, however it seems like it shouldn't too hard to change it to be explicitly written this way by moving some part of `install_file` to `init_install`. What do you think of doing that, and then defining one memoized function to compute the list of all installation entries and calling this memoized function from the 3 places where we need it?

I'm also a bit worried by all the calls to `Rules.file_rule`. It seems to me that after the proposed refactoring, we could build a map of `dir -> entries` in one pass and use that to produce the symlinks. What do you think?

**Comment by diml** [comment]
We discussed it offline and decided to leave the part about `Rules.file_rule` for later

File: src/gen_rules.ml [hunk 1]
```diff
@@ -348,8 +344,27 @@ let gen ~contexts
...
+      | Install ctx ->
+        Option.map (String.Map.find sctxs ctx) ~f:(fun sctx ->
+          (fun ~dir _ ->
+             Install_rules.gen_rules sctx ~dir:(Path.as_in_build_dir_exn dir);
+             Build_system.These String.Set.empty))
+      | Context ctx ->
+        String.Map.find generators ctx);
 
+  String.Map.iter map ~f:(fun (module M : Gen) ->
+    Build_system.handle_add_rule_effects (fun () ->
```
Comment (rgrinberg) [inline src/gen_rules.ml]: Once we port odoc to use `Rules`, we will not need this, right?

File: src/build_system.ml [hunk 2]
```diff
@@ -544,23 +544,36 @@ module Rule_fn = struct
...
+       (dir:Path.t -> string list -> extra_sub_directories_to_keep) option) Fdecl.t
   ; mutable load_dir_stack : Path.t list
   ; (* Set of directories under _build that have at least one rule and
        all their ancestors. *)
     mutable build_dirs_to_keep : Path.Set.t
   ; mutable prefix : (unit, unit) Build.t option
   ; hook : hook -> unit
   ; (* Package files are part of *)
-    packages : Package.Name.t Path.Table.t
+    packages : (Path.t -> Package.Name.t list) Fdecl.t
```
Comment (rgrinberg) [inline src/build_system.ml]: I think it would be a bit more clear if this returned `Package.Name.Set.t` instead.

File: src/gen_rules.ml (hunk 1 above)
Comment (aalekseyev) [reply src/gen_rules.ml]: Yeah, I think so. This changes somewhat in #2089 and I hope will go away at some point.

File: src/build_system.ml (hunk 2 above)
Comment (aalekseyev) [reply src/build_system.ml]: I made this change and I'm merging the PR now.

**Comment by diml** [comment]
@aalekseyev I pushed one last commit to deforest a bit the construction of the map of set

**Comment by aalekseyev** [comment]
@diml thanks

**Pull request merged**